src/fmt/format.go

```diff
@@ -552,10 +553,20 @@ func (f *fmt) fmtFloat(v float64, size int, verb rune, prec int) {
 				}
 				fallthrough
 			default:
-				digits--
+				if num[i] != '0' {
+					sawNonzeroDigit = true
+				}
+				// Count significant digits after saw the first non-zero digit.
+				if sawNonzeroDigit {
+					digits--
+				}
 			}
 		}
 		if !hasDecimalPoint {
+			// decimal 0 should contribute once to digits.
+			if num[1] == '0' && len(num) == 2 {
```

Comment:
nit: It doesn't really matter, but I would have written it as
```
if len(num) == 2 && num[1] == '0' {
    digits--
}
```
Mostly because it asserts the length is higher than the index being accessed.

Reply:
thanks for your advice!